src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -1490,530 +71,116 @@ internal static unsafe uint IUnknown_Release(IntPtr pThis)
-                // so that if it is ever expanded, we are not impacted by that during
-                // enumeration.
-                _buckets = set._buckets;
-                Reset();
+                ManagedObjectWrapper* wrapper = ComInterfaceDispatch.ToManagedObjectWrapper((ComInterfaceDispatch*)pThis);
+                return wrapper->Peg();
             }
 
-            public GCHandle Current
+            [UnmanagedCallersOnly]
+            internal static unsafe uint IReferenceTrackerTarget_Unpeg(IntPtr pThis)
             {
-                get
-                {
-                    if (_currentEntry == null)
-                    {
-                        throw new InvalidOperationException("InvalidOperation_EnumOpCantHappen");
-                    }
-
-                    return _currentEntry.m_value;
-                }
+                ManagedObjectWrapper* wrapper = ComInterfaceDispatch.ToManagedObjectWrapper((ComInterfaceDispatch*)pThis);
+                return wrapper->Unpeg();
             }
 
-            object IEnumerator.Current => Current;
-
-            public void Dispose()
+            [UnmanagedCallersOnly]
+            internal static unsafe int ITaggedImpl_IsCurrentVersion(IntPtr pThis, IntPtr version)
             {
+                return version == (IntPtr)(delegate* unmanaged<IntPtr, IntPtr, int>)&ITaggedImpl_IsCurrentVersion
+                    ? HResults.S_OK
+                    : HResults.E_FAIL;
             }
 
-            public bool MoveNext()
+            static unsafe VtableImplementations()
             {
-                if (_currentEntry != null)
-                {
-                    _currentEntry = _currentEntry.m_next;
-                }
-
-                if (_currentEntry == null)
-                {
-                    // Certain buckets might be empty, so loop until we find
-                    // one with an entry.
-                    while (++_currentIdx != _buckets.Length)
-                    {
-                        _currentEntry = _buckets[_currentIdx];
-                        if (_currentEntry != null)
-                        {
-                            return true;
-                        }
-                    }
+                // Use the "pre-inited vtable" pattern to ensure that ILC can pre-compile these vtables.
+                GetIUnknownImpl(
+                    fpQueryInterface: out *(nint*)&((IUnknownVftbl*)Unsafe.AsPointer(ref IUnknown))->QueryInterface,
+                    fpAddRef: out *(nint*)&((IUnknownVftbl*)Unsafe.AsPointer(ref IUnknown))->AddRef,
+                    fpRelease: out *(nint*)&((IUnknownVftbl*)Unsafe.AsPointer(ref IUnknown))->Release);
 
-                    return false;
-                }
-
-                return true;
-            }
+                IReferenceTrackerTarget.IUnknownVftbl.QueryInterface = (delegate* unmanaged<IntPtr, Guid*, IntPtr*, int>)&IReferenceTrackerTarget_QueryInterface;
```

Comment:
The ILC optimization is very sensitive to the exact IL pattern used, and this cctor is doing some slightly tricky things (eg. the `IReferenceTrackerTarget` is initialized by manually setting QI and then calling `GetIUnknownImpl` to set slots 1-2, and ignoring the QI result. Can we add a test to verify this type is pre-initialized, so we can make sure (1) it works correctly, and (2) that it doesn't accidentally regress in the future? We should be able to copy what Michal did in #114354, just adding a new case in `Preinitialization.cs`.